test/unit/org/apache/cassandra/index/sai/cql/IndexCompressionTest.java

```diff
@@ -0,0 +1,179 @@
+import org.apache.cassandra.index.sai.disk.format.Version;
+import org.apache.cassandra.io.compress.CompressionMetadata;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.File;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+public class IndexCompressionTest extends SAITester
+{
+    @Before
+    public void setup()
+    {
+        if (!Version.latest().onOrAfter(Version.EC))
+            SAIUtil.setLatestVersion(Version.EC);
+    }
+
+    @Test
+    public void testCannotUseCompressionInOlderVersions()
+    {
+        SAIUtil.setLatestVersion(Version.EB);
+        createTable("CREATE TABLE %s (pk int, c text, val text, PRIMARY KEY(pk, c))");
+        var exception = assertThrows(InvalidRequestException.class,
+                                     () -> createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex' WITH options = { 'key_compression': '{\"class\": \"LZ4Compressor\"}' };"));
+        assertTrue(exception.getMessage().contains("Cannot create compressed storage-attached index"));
+        assertTrue(exception.getMessage().contains("ec"));
+        assertTrue(exception.getMessage().contains("eb"));
+    }
+
+    @Test
+    public void testKeyCompression()
+    {
+        createTable("CREATE TABLE %s (pk int, c text, val text, PRIMARY KEY(pk, c))");
+        String indexName = createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex' WITH options = { 'key_compression': '{\"class\": \"LZ4Compressor\"}' };");
+        for (int i = 0; i < 1000; i++)
+            execute("INSERT INTO %s(pk, c, val) VALUES (?, ?, ?)", i, "key", "value" + i);
+
+        flush();
+
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value0'"), 1);
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value5'"), 1);
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value999'"), 1);
+
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        StorageAttachedIndex sai = (StorageAttachedIndex) cfs.indexManager.getIndexByName(indexName);
+        IndexContext context = sai.getIndexContext();
+
+        for (SSTableReader sstable : cfs.getLiveSSTables())
+        {
+            IndexDescriptor descriptor = IndexDescriptor.load(sstable, Set.of(context));
+            assertCompressed(descriptor.perSSTableComponents().get(IndexComponentType.PRIMARY_KEY_TRIE));
+            assertCompressed(descriptor.perSSTableComponents().get(IndexComponentType.PRIMARY_KEY_BLOCKS));
+        }
+    }
+
+    @Test
+    public void testKeyCompressionMustMatchOnAllIndexes()
+    {
+        // Check if we reject creating an index with a different key compression than the one already created.
+        // This is beacuse both indexes share the same primary key map, and it can be compressed in one way only.
```

Comment:
This comment is very nice. Maybe we want to add it to `StorageAttachedIndex.validateOptions` too, at line 325, where key compression is checked?